libvast/src/arrow_table_slice.cpp

```diff
@@ -1141,4 +1146,171 @@ template class arrow_table_slice<fbs::table_slice::arrow::v0>;
 template class arrow_table_slice<fbs::table_slice::arrow::v1>;
 template class arrow_table_slice<fbs::table_slice::arrow::experimental>;
 
+// -- utility functions --------------------------------------------------------
+
+namespace {
+
+std::shared_ptr<arrow::Array>
+convert_subnet_array(const arrow::FixedSizeBinaryArray& arr) {
+  auto* pool = arrow::default_memory_pool();
+  auto subnet_builder = subnet_type::make_arrow_builder(pool);
+  auto& address_builder = subnet_builder->address_builder();
+  auto& length_builder = subnet_builder->length_builder();
+  if (!address_builder.Resize(arr.length()).ok()
+      || !length_builder.Resize(arr.length()).ok()
+      || !subnet_builder->Resize(arr.length()).ok())
+    die("unable to resize builders to target capacity");
+  for (int row = 0; row < arr.length(); ++row) {
+    if (arr.IsNull(row)) {
+      if (!subnet_builder->AppendNull().ok())
+        die("append null value to subnet builder failed");
+    } else {
+      auto span = std::span<const uint8_t, 17>{arr.GetValue(row), 17};
+      if (!(subnet_builder->Append().ok()
+            && address_builder.Append(arr.GetValue(row)).ok()
+            && length_builder.Append(span[16]).ok()))
+        die("rebuilding of subnet struct array failed");
+    }
+  }
+  if (auto result = subnet_builder->Finish(); result.ok())
+    return *result;
+  die("failed to finish Arrow subnet column builder");
```

Comment:
With how many things can go wrong in this function, should we consider making it return an `arrow::Result` or `caf::expected` instead?

Reply:
I made it return an `arrow::Status` because that's what all the potential error paths from this function return anyway, so I could just forward them.